### PR TITLE
[flutter_tools] document flutter root initialization

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Moves the flutter root initialization to a static method on the cache. This is a small step towards making this functionality non-static and instead injected like normal members - however, completely removing all of the static-ness at once was too large of a change.

Instead document and add unit tests and change existing code as little as possible.

https://github.com/flutter/flutter/issues/47161

### CODE REVIEW

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart [hunk 1]
```diff
@@ -156,43 +156,6 @@ class FlutterCommandRunner extends CommandRunner<void> {
...
-      final String script = globals.platform.script.toFilePath();
-      if (globals.fs.path.basename(script) == kSnapshotFileName) {
-        return globals.fs.path.dirname(globals.fs.path.dirname(globals.fs.path.dirname(script)));
-      }
-      if (globals.fs.path.basename(script) == kFlutterToolsScriptFileName) {
-        return globals.fs.path.dirname(globals.fs.path.dirname(globals.fs.path.dirname(globals.fs.path.dirname(script))));
-      }
-
-      // If run from a bare script within the repo.
-      if (script.contains('flutter/packages/')) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart:182]: Could not find any usage of these

File: packages/flutter_tools/lib/src/cache.dart [hunk 2]
```diff
@@ -215,6 +217,69 @@ class Cache {
...
+  ///   6. The current directory
+  ///
+  /// If an exception is thrown during any of these checks, an error message is
+  /// printed and `.` is returned by default (6).
+  static String defaultFlutterRoot({
+    @required Platform platform,
+    @required FileSystem fileSystem,
+    @required UserMessages userMessages,
+  }) {
+    String process(String path) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/cache.dart:246]: Maybe `normalize` instead of `process`?

**Review by zanderso — APPROVED** [review]
lgtm w/ nit

File: packages/flutter_tools/lib/src/cache.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/cache.dart]: done

**Pull request merged**